cmd/soroban-cli/src/events.rs

```diff
@@ -0,0 +1,304 @@
+                low: self.start_ledger,
+                high: self.end_ledger,
+            });
+        }
+
+        for raw_contract_id in &self.contract_ids {
+            // We parse the contract IDs to ensure they're the correct format,
+            // but since we'll be passing them as-is to the RPC server anyway,
+            // we disregard the return value.
+            utils::id_from_str::<32>(raw_contract_id).map_err(|e| Error::InvalidContractId {
+                contract_id: raw_contract_id.clone(),
+                error: e,
+            })?;
+        }
+
+        let mut events: Vec<Event> = Vec::new();
+        if let Some(rpc_url) = self.rpc_url.as_ref() {
+            let client = Client::new(rpc_url);
+            let rpc_event = client.get_events(&self.contract_ids, &self.topics).await?;
+            events = rpc_event.events;
+        } else if let Some(path) = self.events_file.as_ref() {
+            if !path.exists() {
+                return Err(Error::InvalidFile {
+                    path: path.to_str().unwrap().to_string(),
+                });
+            }
+
+            // Read the JSON events from disk and find the ones that match the
+            // contract ID filter(s) that were passed in.
+            events.extend(
+                snapshot::read_events(path)
+                    .map_err(|err| Error::CannotReadFile {
+                        path: path.to_str().unwrap().to_string(),
+                        error: err.to_string(),
+                    })?
+                    .iter()
+                    // The ledger range isn't optional, so filter on that first.
+                    .filter(|evt| {
+                        match evt.ledger.parse::<u32>() {
+                            Ok(seq) => seq >= self.start_ledger && seq <= self.end_ledger,
+                            Err(e) => {
+                                eprintln!(
+                                    "error parsing key 'ledger' ('{:?}'): {:#?}",
+                                    evt.ledger, e
+                                );
+                                false // eat error
+                            }
+                        }
+                    })
+                    .filter(|evt| {
+                        // Contract ID filter(s) are optional, so we should
+                        // render all events if they're omitted.
+                        self.contract_ids.is_empty()
+                            || self.contract_ids.iter().any(|id| *id == evt.contract_id)
+                    })
+                    .filter(|evt| {
+                        // Like before, no topic means pass everything through.
+                        self.topics.is_empty()
+                            || self
+                                .topics
+                                .iter()
+                                // quadratic but both are <= 5 long
+                                .any(|t| evt.topic.iter().any(|t2| *t == *t2))
```

Comment:
I'm not sure this topic matching logic is right. It's missing wildcards for one, and seems like it's doing character comparison?

Reply:
Might be nice to pull this into a function and do some unit tests. You could have a look at the go unit tests for some example cases.
https://github.com/stellar/soroban-tools/blob/main/cmd/soroban-rpc/internal/methods/get_events_test.go#L21

Reply:
Nice, great idea - I did this haphazardly to get a "this is what it would look like" out there

Reply:
Improved this but untested yet

Reply:
Tested now w/ unit tests from RPC